provider/oci/environ.go

```diff
@@ -223,42 +378,313 @@ func (e *Environ) Destroy(ctx envcontext.ProviderCallContext) error {
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	zone := zones[0].Name()
+	network := networks[zone][0]
+	// refresh the global image cache
+	// this only hits the API every 30 minutes, otherwise just retrieves
+	// from cache
+	imgCache, err := refreshImageCache(e.Compute, e.ecfg().compartmentID())
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	logger.Debugf("Image cache contains: %v", imgCache)
+	series := args.Tools.OneSeries()
+	arches := args.Tools.Arches()
+
+	types := imgCache.SupportedShapes(series)
+
+	defaultType := string(VirtualMachine)
+	if args.Constraints.VirtType == nil {
+		args.Constraints.VirtType = &defaultType
+	}
+
+	// check if we find an image that is compliant with the
+	// constraints provided in the oracle cloud account
+	args.ImageMetadata = imgCache.ImageMetadata(series, *args.Constraints.VirtType)
+
+	spec, image, err := findInstanceSpec(
+		args.ImageMetadata,
+		types,
+		&instances.InstanceConstraint{
+			Series:      series,
+			Arches:      arches,
+			Constraints: args.Constraints,
+		},
+	)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	tools, err := args.Tools.Match(tools.Filter{Arch: spec.Image.Arch})
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	logger.Tracef("agent binaries: %v", tools)
+	if err = args.InstanceConfig.SetTools(tools); err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	if err = instancecfg.FinishInstanceConfig(
+		args.InstanceConfig,
+		e.Config(),
+	); err != nil {
+		return nil, errors.Trace(err)
+	}
+	hostname, err := e.namespace.Hostname(args.InstanceConfig.MachineId)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	tags := args.InstanceConfig.Tags
+
+	var apiPort int
+	var desiredStatus ociCore.InstanceLifecycleStateEnum
+	// Wait for controller to actually be running
+	if args.InstanceConfig.Controller != nil {
+		apiPort = args.InstanceConfig.Controller.Config.APIPort()
+		desiredStatus = ociCore.InstanceLifecycleStateRunning
+	} else {
+		desiredStatus = ociCore.InstanceLifecycleStateProvisioning
+	}
+
+	cloudcfg, err := e.getCloudInitConfig(series, apiPort)
+	if err != nil {
+		return nil, errors.Annotate(err, "cannot create cloudinit template")
+	}
+
+	// compose userdata with the cloud config template
+	logger.Debugf("Composing userdata")
+	userData, err := providerinit.ComposeUserData(
+		args.InstanceConfig,
+		cloudcfg,
+		OCIRenderer{},
+	)
+	if err != nil {
+		return nil, errors.Annotate(err, "cannot make user data")
+	}
+
+	var rootDiskSize int
+	if args.Constraints.RootDisk != nil {
+		rootDiskSize = int(*args.Constraints.RootDisk)
+		if int(*args.Constraints.RootDisk) < MinVolumeSizeMB {
+			rootDiskSize = MinVolumeSizeMB / 1024
+		} else if int(*args.Constraints.RootDisk) > MinVolumeSizeMB {
+			rootDiskSize = MaxVolumeSizeMB / 1024
+		}
+	} else {
+		rootDiskSize = MinVolumeSizeMB / 1024
+	}
+
+	// NOTE(gsamfira): Should we make this configurable?
+	// TODO(gsamfira): select Availability Domain and subnet ID
```

Comment:
what might be configurable here?
We looked up the image, root disk was passed in, compartment id was given.
"zone" is something we'll probably eventually want to support, but doesn't have to be there day 1,

I'm not sure if this comment is no longer relevant, or just needs to be expanded for what we may want to add configuration in the future.